src/audio/pipewire/SDL_pipewire.c

```diff
@@ -0,0 +1,1088 @@
+                                                                 SPA_AUDIO_CHANNEL_SL,  SPA_AUDIO_CHANNEL_SR };
+
+#define COPY_CHANNEL_MAP(c) SDL_memcpy(info->position, PIPEWIRE_channel_map_##c, sizeof(PIPEWIRE_channel_map_##c))
+
+static void
+initialize_spa_info(const SDL_AudioSpec *spec, struct spa_audio_info_raw *info)
+{
+  info->channels = spec->channels;
+  info->rate     = spec->freq;
+
+  switch (spec->channels) {
+  case 1:
+    COPY_CHANNEL_MAP(1);
+    break;
+  case 2:
+    COPY_CHANNEL_MAP(2);
+    break;
+  case 3:
+    COPY_CHANNEL_MAP(3);
+    break;
+  case 4:
+    COPY_CHANNEL_MAP(4);
+    break;
+  case 5:
+    COPY_CHANNEL_MAP(5);
+    break;
+  case 6:
+    COPY_CHANNEL_MAP(6);
+    break;
+  case 7:
+    COPY_CHANNEL_MAP(7);
+    break;
+  case 8:
+    COPY_CHANNEL_MAP(8);
+    break;
+  }
+
+  /* Pipewire natively supports all of SDL's sample formats */
+  switch (spec->format) {
+  case AUDIO_U8:
+    info->format = SPA_AUDIO_FORMAT_U8;
+    break;
+  case AUDIO_S8:
+    info->format = SPA_AUDIO_FORMAT_S8;
+    break;
+  case AUDIO_U16LSB:
+    info->format = SPA_AUDIO_FORMAT_U16_LE;
+    break;
+  case AUDIO_S16LSB:
+    info->format = SPA_AUDIO_FORMAT_S16_LE;
+    break;
+  case AUDIO_U16MSB:
+    info->format = SPA_AUDIO_FORMAT_U16_BE;
+    break;
+  case AUDIO_S16MSB:
+    info->format = SPA_AUDIO_FORMAT_S16_BE;
+    break;
+  case AUDIO_S32LSB:
+    info->format = SPA_AUDIO_FORMAT_S32_LE;
+    break;
+  case AUDIO_S32MSB:
+    info->format = SPA_AUDIO_FORMAT_S32_BE;
+    break;
+  case AUDIO_F32LSB:
+    info->format = SPA_AUDIO_FORMAT_F32_LE;
+    break;
+  case AUDIO_F32MSB:
+    info->format = SPA_AUDIO_FORMAT_F32_BE;
+    break;
+  }
+}
+
+static void
+output_callback(void *data)
+{
+  struct pw_buffer * pw_buf;
+  struct spa_buffer *spa_buf;
+  Uint8 *            dst;
+
+  _THIS                    = (SDL_AudioDevice *)data;
+  struct pw_stream *stream = this->hidden->stream;
+
+  /* Shutting down, don't do anything */
+  if (SDL_AtomicGet(&this->shutdown)) {
+    return;
+  }
+
+  /* See if a buffer is available */
+  if ((pw_buf = PIPEWIRE_pw_stream_dequeue_buffer(stream)) == NULL) {
+    return;
+  }
+
+  spa_buf = pw_buf->buffer;
+
+  if (spa_buf->datas[0].data == NULL) {
+    return;
+  }
+
+  /*
+   * If the device is disabled, write silence to the stream buffer
+   * and run the callback with the work buffer to keep the callback
+   * firing regularly in case the audio is being used as a timer.
+   */
+  if (SDL_AtomicGet(&this->enabled)) {
+    dst = spa_buf->datas[0].data;
+  } else {
+    dst = this->work_buffer;
+    SDL_memset(spa_buf->datas[0].data, this->spec.silence, this->spec.size);
+  }
+
+  if (!SDL_AtomicGet(&this->paused)) {
+    if (!this->stream) {
+      SDL_LockMutex(this->mixer_lock);
+      this->callbackspec.callback(this->spec.userdata, dst, this->callbackspec.size);
```

Comment:
```suggestion
      this->callbackspec.callback(this->callbackspec.userdata, dst, this->callbackspec.size);
```
This is broken in case the app uses SDL_QueueAudio() otherwise. Example app: https://gist.github.com/Oschowa/1f8f8b17ad974bf0d6a3346cb0c83249

Reply:
Yep, that should be callbackspec.userdata. Thanks for catching this!